packages/client/src/schedule-types.ts

```diff
@@ -0,0 +1,830 @@
+   * running.
+   *
+   * @default {@link ScheduleOverlapPolicy.SKIP}
+   */
+  overlap: ScheduleOverlapPolicy;
+
+  /**
+   * The Temporal Server might be down or unavailable at the time when a Schedule should take an Action. When the Server
+   * comes back up, `catchupWindow` controls which missed Actions should be taken at that point. The default is one
+   * minute, which means that the Schedule attempts to take any Actions that wouldn't be more than one minute late. It
+   * takes those Actions according to the {@link ScheduleOverlapPolicy}. An outage that lasts longer than the Catchup
+   * Window could lead to missed Actions. (But you can always {@link ScheduleHandle.backfill}.)
+   *
+   * @default 1 minute
+   * @format number of milliseconds
+   */
+  catchupWindow: number | undefined;
+
+  /**
+   * When an Action times out or reaches the end of its Retry Policy, {@link pause}.
+   *
+   * With {@link ScheduleOverlapPolicy.ALLOW_ALL}, this pause might not apply to the next Action, because the next Action
+   * might have already started previous to the failed one finishing. Pausing applies only to Actions that are scheduled
+   * to start after the failed one finishes.
+   *
+   * @default false
+   */
+  pauseOnFailure: boolean;
+
+  /**
+   * The Actions remaining in this Schedule. Once this number hits `0`, no further Actions are taken (unless {@link
+   * ScheduleHandle.trigger} is called).
+   *
+   * @default undefined (unlimited)
+   */
+  remainingActions?: number;
+
+  /** Number of Actions taken so far. */
+  numActionsTaken: number;
+
+  /** Number of times a scheduled Action was skipped due to missing the catchup window. */
+  numActionsMissedCatchupWindow: number;
+
+  /** Number of Actions skipped due to overlap. */
+  numActionsSkippedOverlap: number;
+
+  /**
+   * Currently-running workflows started by this schedule. (There might be
+   * more than one if the overlap policy allows overlaps.)
+   */
+  runningWorkflows: WorkflowExecutionWithFirstExecutionRunId[];
+
+  createdAt: Date;
+  lastUpdatedAt: Date | undefined;
+
+  raw: temporal.api.workflowservice.v1.IDescribeScheduleResponse;
+};
+
+export type UpdatedSchedule<Action extends Workflow> = Pick<
+  ScheduleOptions<Action>,
+  'spec' | 'action' | 'overlap' | 'catchupWindow' | 'pauseOnFailure' | 'note' | 'paused'
+>;
+
+/**
+ * Make all properties optional.
+ *
+ * If original Schedule is deleted, okay to use same `id`.
+ */
+export type ScheduleOptionsOverrides<Action extends Workflow> = Partial<ScheduleOptions<Action>>;
+
+export interface WorkflowExecutionWithFirstExecutionRunId {
+  workflowId: string;
+
+  /**
+   * The Run Id of the original execution that was started by the Schedule. If the Workflow retried, did
+   * Continue-As-New, or was Reset, the following runs would have different Run Ids.
+   */
+  firstExecutionRunId: string;
+}
+
+export interface ScheduleAction {
```

Comment:
I understand that we might want to tweak some names at this level, but it does seem confusing to have a `ScheduleAction` that corresponds to a `ScheduleActionResult` in the proto, while `ScheduleAction` in the proto is something else

Reply:
Definitely agreed. That name is also confusing in regard to other types in the TypeScript's schedule API.

I'm considering this:
```
export interface ScheduleExecutionStartWorkflowActionResult {
  /** Time that the Action was scheduled for, including jitter */
  scheduledAt: Date;

  /** Time that the Action was actually taken */
  takenAt: Date;

  /** The Action that was taken */
  action: ScheduleExecutionActionResult;
}

export type ScheduleExecutionActionResult = ScheduleExecutionStartWorkflowActionResult;

export interface ScheduleExecutionStartWorkflowActionResult {
  type: 'startWorkflow';
  workflow: WorkflowExecutionWithFirstExecutionRunId;
}
```

Reply:
This API hides all protos, right? So I don't see the issue with name conflicts. And generally prefer shorter terms, like:

```ts
export interface ScheduleAction {
  /** Time that the Action was scheduled for, including jitter */
  scheduledAt: Date;

  /** Time that the Action was actually taken */
  takenAt: Date;

  /** The result of the Action that was taken */
  result: ScheduleActionResult;
}

export type ScheduleActionResult = StartWorkflowActionResult;

export interface StartWorkflowActionResult {
  type: 'startWorkflow';
  workflow: WorkflowExecutionWithFirstExecutionRunId;
}
```

Reply:
@lorensr I obviously prefer sorter names.

The problem here is that Describe, List, Create/Update, CompiledCreate/CompiledUpdate, recentActions, and runningActions all have distinct variants of the `ScheduleAction*` types. At this moment, we could deal with this by `Omit<>`ing or `Pick<>`ing some fields, but that would not work once we add other types of actions.